lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,269 @@
+            earned_graded (float)
+            possible_graded (float)
+            visible_blocks (iterable of BlockRecord)
+        """
+        visible_blocks = kwargs.pop('visible_blocks')
+        if not isinstance(visible_blocks, BlockRecordList):
+            visible_blocks = BlockRecordList(visible_blocks)
+        try:
+            visible_blocks_model = VisibleBlocks.objects.create_from_blockrecords(blocks=visible_blocks)
+        except IntegrityError:
+            # If an integrity error occurs, the VisibleBlocks model we want to
+            # create already exists.  Use the hash for our current list of
+            # blocks as the foreign key.
+            visible_blocks_hash = visible_blocks.to_hash()
+        else:
+            visible_blocks_hash = visible_blocks_model.hashed
+
+        grade = self.model(
+            course_id=kwargs['usage_key'].course_key,
+            visible_blocks_id=visible_blocks_hash,
+            **kwargs
+        )
+        grade.full_clean()
+        grade.save()
+        return grade
+
+
+class PersistentSubsectionGrade(TimeStampedModel):
+    """
+    A django model tracking persistent grades at the subsection level.
+    """
+
+    class Meta(object):
+        unique_together = (('user_id', 'course_id', 'usage_key'), )
+
+    # primary key will need to be large for this table
+    id = UnsignedBigIntAutoField(primary_key=True)  # pylint: disable=invalid-name
+
+    # uniquely identify this particular grade object
+    user_id = models.IntegerField(blank=False)
+    course_id = CourseKeyField(blank=False, max_length=255)
+    usage_key = UsageKeyField(blank=False, max_length=255)
+
+    # Information relating to the state of content when grade was calculated
+    subtree_edited_date = models.DateTimeField('last content edit timestamp', blank=False)
+    course_version = models.CharField('guid of latest course version', blank=False, max_length=255)
+
+    # earned/possible refers to the number of points achieved and available to achieve.
+    # graded refers to the subset of all problems that are marked as being graded.
+    earned_all = models.FloatField(blank=False)
+    possible_all = models.FloatField(blank=False)
+    earned_graded = models.FloatField(blank=False)
+    possible_graded = models.FloatField(blank=False)
+
+    # track which blocks were visible at the time of grade calculation
+    visible_blocks = models.ForeignKey(VisibleBlocks, db_column='visible_blocks_hash', to_field='hashed')
+
+    # use custom manager
+    objects = PersistentSubsectionGradeQuerySet.as_manager()
+
+    def __unicode__(self):
+        """
+        Returns a string representation of this model.
+        """
+        return "PersistentSubsectionGrade user:{}, subsection {}. {}/{} graded, {}/{} all".format(
+            self.user_id,
+            self.usage_key,
+            self.earned_graded,
+            self.possible_graded,
+            self.earned_all,
+            self.possible_all,
+        )
```

Comment:
This super descriptive `unicode` method came in handy for me while debugging tests, big 👍 here